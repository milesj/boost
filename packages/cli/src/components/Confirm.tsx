/* eslint-disable no-console */

import React, { useCallback, useEffect, useState } from 'react';
import { useFocus } from 'ink';
import { msg } from '../translate';
import { Prompt, PromptProps } from './internal/Prompt';
import { Selected } from './internal/Selected';
import { Style } from './Style';

export interface ConfirmProps extends Omit<PromptProps<boolean>, 'validate'> {
	/** Error message to display when an invalid character is pressed. Defaults to a custom message. */
	invalidError?: string;
	/** Character that triggers a falsy state when pressed. Defaults to "N". */
	no?: string;
	/** Character that triggers a truthy state when pressed. Defaults to "y". */
	yes?: string;
}

/**
 * A React component that renders a yes/no confirmation prompt.
 */
export function Confirm(props: ConfirmProps) {
	const { invalidError, onSubmit, no = 'N', yes = 'y', ...restProps } = props;
	const [value, setValue] = useState<boolean | null>(null);
	const { isFocused } = useFocus({ autoFocus: true });

	useEffect(() => {
		if (no.length !== 1) {
			console.error(`Prop "no" must be a single character.`);
		} else if (yes.length !== 1) {
			console.error(`Prop "yes" must be a single character.`);
		}
	}, [no, yes]);

	const handleInput = useCallback(
		(input: string) => {
			if (input === no) {
				setValue(false);
				onSubmit(false);
			} else if (input === yes) {
				setValue(true);
				onSubmit(true);
			} else {
				setValue(null);

				throw new Error(invalidError ?? msg('prompt:confirmInvalidValue', { no, yes }));
			}

			// Trigger submit
			return true;
		},
		[invalidError, no, onSubmit, yes],
	);

	return (
		<Prompt<boolean>
			{...restProps}
			afterLabel={
				value === null ? <Style type="muted">{`(${yes}/${no})`}</Style> : <Selected value={value} />
			}
			focused={isFocused}
			value={value}
			onInput={handleInput}
		/>
	);
}
