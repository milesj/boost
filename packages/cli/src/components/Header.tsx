import React from 'react';
import { Box } from 'ink';
import { SPACING_ROW } from '../constants';
import { StyleType } from '../types';
import { Style } from './Style';

export interface HeaderProps {
	/** Text to display for the label. */
	label: string;
	/** The bottom margin. Defaults to `1`. */
	marginTop?: number;
	/** The top margin. Defaults to `1`. */
	marginBottom?: number;
	/** Customize the background color using Style. Defaults to normal text. */
	type?: StyleType | 'none';
}

/**
 * A React component that renders a header/title within a terminal menu.
 */
export function Header(props: HeaderProps) {
	const { label, marginTop = SPACING_ROW, marginBottom = SPACING_ROW, type = 'default' } = props;

	return (
		<Box marginBottom={marginBottom} marginTop={marginTop}>
			<Style bold inverted type={type}>
				{` ${label.toLocaleUpperCase()} `}
			</Style>
		</Box>
	);
}
