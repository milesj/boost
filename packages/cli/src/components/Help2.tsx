import React from 'react';
import { Text } from 'ink';

export function Help() {
  return <Box flexDirection="column"></Box>;
}
