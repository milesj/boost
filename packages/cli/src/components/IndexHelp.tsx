import React from 'react';
import { Box, Text } from 'ink';
import { SPACING_ROW } from '../constants';
import { ProgramOptions } from '../types';
import { Style } from './Style';

export interface IndexHelpProps extends ProgramOptions {
	children?: React.ReactNode;
}

/**
 * A React component that renders a help menu for the entire program.
 */
export function IndexHelp(props: IndexHelpProps) {
	const { banner, bin, children, footer, header, name, version } = props;

	return (
		<Box flexDirection="column" marginBottom={1}>
			{!!banner && (
				<Box>
					<Text>{banner}</Text>
				</Box>
			)}

			<Box marginTop={SPACING_ROW}>
				<Text>
					{`${name} v${version}`} <Style type="muted">{bin}</Style>
				</Text>
			</Box>

			{!!header && (
				<Box marginTop={SPACING_ROW}>
					<Text>{header}</Text>
				</Box>
			)}

			{children}

			{!!footer && (
				<Box marginTop={SPACING_ROW}>
					<Text>{footer}</Text>
				</Box>
			)}
		</Box>
	);
}
