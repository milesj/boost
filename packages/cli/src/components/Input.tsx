import React, { useCallback, useState } from 'react';
import { useFocus } from 'ink';
import { Cursor } from './internal/Cursor';
import { Prompt, PromptProps } from './internal/Prompt';
import { Style } from './Style';

export interface InputProps extends PromptProps<string> {
	/** A default value. If none provided, will use an empty state. */
	defaultValue?: string;
	/** Hide the cursor in the terminal. Will remove the background color, but still functions. */
	hideCursor?: boolean;
	/** Mask to replace every inputted character with. */
	mask?: string;
	/** Callback triggered when the value changes. */
	onChange?: (value: string) => void;
	/** Custom string to display when the value is empty and non-dirty. */
	placeholder?: string;
}

/**
 * A React component that renders an input field, allowing the user to enter information.
 */
export function Input(props: InputProps) {
	const {
		defaultValue = '',
		hideCursor,
		mask,
		onChange,
		onSubmit,
		placeholder,
		...restProps
	} = props;
	const [value, setValue] = useState(defaultValue);
	const [cursorPosition, setCursorPosition] = useState(0);
	const [isDirty, setDirty] = useState(false);
	const { isFocused } = useFocus({ autoFocus: true });

	const handleReturn = useCallback(() => {
		onSubmit(value.trim());

		// Trigger submit
		return true;
	}, [onSubmit, value]);

	// Remove characters
	const handleBackspace = useCallback(() => {
		if (!cursorPosition || !value) {
			return;
		}

		const nextValue =
			cursorPosition >= value.length
				? value.slice(0, -1)
				: value.slice(0, cursorPosition - 1) + value.slice(cursorPosition);

		setCursorPosition(cursorPosition - 1);
		setValue(nextValue);
		onChange?.(nextValue);
	}, [cursorPosition, onChange, value]);

	// Add characters
	const handleInput = useCallback(
		(input: string) => {
			const nextValue =
				cursorPosition >= value.length
					? value + input
					: value.slice(0, cursorPosition) + input + value.slice(cursorPosition);

			setCursorPosition(cursorPosition + input.length);
			setValue(nextValue);
			setDirty(true);
			onChange?.(nextValue);
		},
		[cursorPosition, onChange, value],
	);

	// Navigation
	const handleKeyUp = useCallback(() => {
		setCursorPosition(0);
	}, []);

	const handleKeyDown = useCallback(() => {
		setCursorPosition(value.length);
	}, [value]);

	const handleKeyLeft = useCallback(() => {
		setCursorPosition((prev) => Math.max(prev - 1, 0));
	}, []);

	const handleKeyRight = useCallback(() => {
		setCursorPosition((prev) => Math.min(prev + 1, value.length));
	}, [value]);

	return (
		<Prompt<string>
			{...restProps}
			afterLabel={
				value === '' && !isDirty && placeholder ? (
					<Style type="muted">{placeholder}</Style>
				) : (
					<Cursor
						focused={isFocused}
						hideCursor={hideCursor}
						position={cursorPosition}
						value={mask === undefined ? value : mask.repeat(value.length)}
					/>
				)
			}
			focused={isFocused}
			value={value}
			onBackspace={handleBackspace}
			onDelete={handleBackspace}
			onInput={handleInput}
			onKeyDown={handleKeyDown}
			onKeyLeft={handleKeyLeft}
			onKeyRight={handleKeyRight}
			onKeyUp={handleKeyUp}
			onReturn={handleReturn}
		/>
	);
}
