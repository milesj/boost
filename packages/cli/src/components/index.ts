export * from './Confirm';
export * from './Failure';
export * from './Header';
export * from './Help';
export * from './HiddenInput';
export * from './IndexHelp';
export * from './Input';
export * from './MultiSelect';
export * from './PasswordInput';
export * from './Select';
export * from './Style';
