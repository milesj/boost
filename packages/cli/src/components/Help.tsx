/* eslint-disable react/destructuring-assignment */

import React from 'react';
import { Box, Text } from 'ink';
import { OptionConfig, OptionConfigMap, ParamConfig, ParamConfigList } from '@boost/args';
import { toArray } from '@boost/common';
import { stripAnsi } from '@boost/terminal';
import { DELIMITER, SPACING_COL, SPACING_COL_WIDE, SPACING_ROW } from '../constants';
import {
	formatCommandCall,
	formatDescription,
	formatType,
	getLongestWidth,
	groupByCategory,
} from '../helpers';
import { msg } from '../translate';
import { Categories, CommandConfig, CommandConfigMap } from '../types';
import { Header } from './Header';
import { Style } from './Style';

export interface HelpProps {
	categories?: Categories;
	config?: CommandConfig;
	commands?: CommandConfigMap;
	delimiter?: string;
	header?: string;
	options?: OptionConfigMap;
	params?: ParamConfigList;
}

export class Help extends React.Component<HelpProps> {
	gatherOptionTags(config: OptionConfig): string[] {
		const tags: string[] = [];

		if (config.count) {
			tags.push(msg('cli:tagCount'));
		}

		if (config.multiple) {
			tags.push(msg('cli:tagMultiple') + (config.arity ? `(${config.arity})` : ''));
		}

		return tags;
	}

	gatherParamTags(config: ParamConfig): string[] {
		const tags: string[] = [];

		if (config.required) {
			tags.push(msg('cli:tagRequired'));
		}

		return tags;
	}

	renderCommands(commands: CommandConfigMap) {
		let pathWidth = 0;

		const items = Object.entries(commands).map(([path, command]) => {
			const pathCall = formatCommandCall(path, command);
			const pathCallStripped = stripAnsi(pathCall);

			if (pathCallStripped.length > pathWidth) {
				pathWidth = pathCallStripped.length;
			}

			return {
				...command,
				name: path,
				path,
				pathCall,
			};
		});

		const categorizedCommands = groupByCategory(items, this.props.categories ?? {});
		const categoryCount = Object.keys(categorizedCommands).length;

		return (
			<>
				<Header label={msg('cli:labelCommands')} />

				{Object.entries(categorizedCommands).map(([key, category], index) => (
					<Box key={key} flexDirection="column">
						{!!category.name && categoryCount > 1 && (
							<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
								<Style bold type="none">
									{category.name}
								</Style>
							</Box>
						)}

						{category.items.map((config) => {
							const desc = formatDescription(config);

							return (
								<Box key={key + config.path} flexDirection="row" paddingLeft={SPACING_COL}>
									<Box flexGrow={0} width={pathWidth + SPACING_COL_WIDE}>
										<Text>{config.pathCall}</Text>
									</Box>

									<Box flexGrow={1}>
										<Text wrap="wrap">{desc}</Text>
									</Box>
								</Box>
							);
						})}
					</Box>
				))}
			</>
		);
	}

	renderOptions(options: OptionConfigMap) {
		let longWidth = 0;
		let shortWidth = 0;

		const items = Object.entries(options).map(([name, option]) => {
			const shortLabel = option.short ? `-${option.short},` : '';
			let longLabel = `--${option.default === true ? 'no-' : ''}${name}`;

			if (option.type !== 'boolean') {
				longLabel += ' ';
				longLabel += formatType(option);
			}

			const longLabelStripped = stripAnsi(longLabel);
			const shortLabelStripped = stripAnsi(shortLabel);

			if (longLabelStripped.length > longWidth) {
				longWidth = longLabelStripped.length;
			}

			if (shortLabelStripped.length > shortWidth) {
				shortWidth = shortLabelStripped.length;
			}

			return {
				...option,
				longLabel,
				name,
				shortLabel,
			};
		});

		const categorizedOptions = groupByCategory(items, this.props.categories ?? {});
		const categoryCount = Object.keys(categorizedOptions).length;
		const showShortColumn = shortWidth > 0;

		return (
			<>
				<Header label={msg('cli:labelOptions')} />

				{Object.entries(categorizedOptions).map(([key, category], index) => (
					<Box key={key} flexDirection="column">
						{!!category.name && categoryCount > 1 && (
							<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
								<Style bold type="none">
									{category.name}
								</Style>
							</Box>
						)}

						{category.items.map((config) => (
							<Box key={key + config.name} flexDirection="row" paddingLeft={SPACING_COL}>
								{showShortColumn && (
									<Box flexGrow={0} width={shortWidth + SPACING_COL}>
										<Text>{config.shortLabel}</Text>
									</Box>
								)}

								<Box flexGrow={0} width={longWidth + SPACING_COL_WIDE}>
									<Text>{config.longLabel}</Text>
								</Box>

								<Box flexGrow={1}>
									<Text wrap="wrap">
										{formatDescription(config, this.gatherOptionTags(config))}
									</Text>
								</Box>
							</Box>
						))}
					</Box>
				))}
			</>
		);
	}

	renderParams(params: ParamConfigList) {
		const labels = params.map((config, index) => `${config.label ?? index} ${formatType(config)}`);
		const labelWidth = getLongestWidth(labels);
		const allowVariadic = this.props.config?.allowVariadicParams;

		return (
			<>
				<Header label={msg('cli:labelParams')} />

				{params.map((config, index) => {
					if (config.hidden) {
						return null;
					}

					const desc = config ? formatDescription(config, this.gatherParamTags(config)) : '';

					return (
						<Box key={`${labels[index]}-${index}`} flexDirection="row" paddingLeft={SPACING_COL}>
							<Box flexGrow={0} width={labelWidth + SPACING_COL_WIDE}>
								<Text>{labels[index]}</Text>
							</Box>

							<Box flexGrow={1}>
								<Text wrap="wrap">{desc}</Text>
							</Box>
						</Box>
					);
				})}

				{allowVariadic && (
					<Box paddingLeft={SPACING_COL}>
						<Text>
							{`â€¦${typeof allowVariadic === 'string' ? allowVariadic : ''}`}{' '}
							{formatType({
								multiple: true,
								type: 'string',
							})}
						</Text>
					</Box>
				)}
			</>
		);
	}

	renderUsage(usage: string[] | string = '') {
		const { delimiter = DELIMITER } = this.props;

		return (
			<>
				<Header label={msg('cli:labelUsage')} />

				{toArray(usage).map((example) => (
					<Box key={example} paddingLeft={SPACING_COL}>
						<Text>{example.startsWith(delimiter) ? example : delimiter + example}</Text>
					</Box>
				))}
			</>
		);
	}

	// eslint-disable-next-line complexity
	render() {
		const { commands, options, header, params, config } = this.props;
		const hasDesc = Boolean(config?.description);
		const hasUsage = Boolean(config?.usage && config.usage.length > 0);
		const hasParams = Boolean(params && params.length > 0);
		const hasCommands = Boolean(commands && Object.keys(commands).length > 0);
		const hasOptions = Boolean(options && Object.keys(options).length > 0);

		return (
			<Box flexDirection="column">
				{(hasDesc || hasUsage) && header && <Header label={header} />}

				{hasDesc && (
					<Box paddingLeft={SPACING_COL}>
						<Text>{formatDescription(config!)}</Text>
					</Box>
				)}

				{hasUsage && this.renderUsage(config?.usage)}

				{hasParams && this.renderParams(params!)}

				{hasCommands && this.renderCommands(commands!)}

				{hasOptions && this.renderOptions(options!)}
			</Box>
		);
	}
}
