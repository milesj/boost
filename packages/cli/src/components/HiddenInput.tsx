import React from 'react';
import { Input, InputProps } from './Input';

export type HiddenInputProps = Omit<InputProps, 'hideCursor' | 'mask'>;

/**
 * A React component that renders an input field that hides its content from the terminal.
 */
export function HiddenInput(props: HiddenInputProps) {
	return <Input {...props} hideCursor mask="" />;
}
