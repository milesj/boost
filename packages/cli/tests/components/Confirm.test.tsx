import React from 'react';
import { render } from 'ink-testing-library';
import { Confirm, ConfirmProps } from '../../src/components/Confirm';

describe('Confirm', () => {
	const props: ConfirmProps = {
		label: 'Confirm?',
		onSubmit() {},
	};

	it('renders label with yes/no by default', () => {
		const { lastFrame } = render(<Confirm {...props} />);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('can customize yes/no', () => {
		const { lastFrame } = render(<Confirm {...props} no="F" yes="t" />);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('console errors if `yes` is invalid', async () => {
		const spy = jest.spyOn(console, 'error').mockImplementation();

		render(<Confirm {...props} yes="" />);

		await delay();

		expect(spy).toHaveBeenCalledWith('Prop "yes" must be a single character.');

		spy.mockRestore();
	});

	it('console errors if `no` is invalid', async () => {
		const spy = jest.spyOn(console, 'error').mockImplementation();

		render(<Confirm {...props} no="" />);

		await delay();

		expect(spy).toHaveBeenCalledWith('Prop "no" must be a single character.');

		spy.mockRestore();
	});

	it('errors for invalid input value', async () => {
		const { lastFrame, stdin } = render(<Confirm {...props} />);

		await delay();
		stdin.write('g');
		await delay();

		expect(lastFrame()).toMatchSnapshot();
	});

	it('errors with custom message', async () => {
		const { lastFrame, stdin } = render(<Confirm {...props} invalidError="Wrong value yo" />);

		await delay();
		stdin.write('g');
		await delay();

		expect(lastFrame()).toMatchSnapshot();
	});

	it('errors for incorrectly cased input value', async () => {
		const { lastFrame, stdin } = render(<Confirm {...props} />);

		await delay();
		stdin.write('Y');
		await delay();

		expect(lastFrame()).toMatchSnapshot();
	});

	it('doesnt call `onSubmit` invalid input value', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Confirm {...props} onSubmit={spy} />);

		await delay();
		stdin.write('g');
		await delay();

		expect(spy).not.toHaveBeenCalled();
	});

	describe('truthy', () => {
		it('renders success state', async () => {
			const { lastFrame, stdin } = render(<Confirm {...props} />);

			await delay();
			stdin.write('y');
			await delay();

			expect(lastFrame()).toMatchSnapshot();
		});

		it('calls `onSubmit`', async () => {
			const spy = jest.fn();
			const { stdin } = render(<Confirm {...props} onSubmit={spy} />);

			await delay();
			stdin.write('y');
			await delay();

			expect(spy).toHaveBeenCalledWith(true);
		});
	});

	describe('falsy', () => {
		it('renders success state', async () => {
			const { lastFrame, stdin } = render(<Confirm {...props} />);

			await delay();
			stdin.write('N');
			await delay();

			expect(lastFrame()).toMatchSnapshot();
		});

		it('calls `onSubmit`', async () => {
			const spy = jest.fn();
			const { stdin } = render(<Confirm {...props} onSubmit={spy} />);

			await delay();
			stdin.write('N');
			await delay();

			expect(spy).toHaveBeenCalledWith(false);
		});
	});
});
