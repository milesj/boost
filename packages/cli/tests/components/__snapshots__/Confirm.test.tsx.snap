// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirm can customize yes/no 1`] = `"[36m?[39m [1mConfirm?[22m [90m(t/F)[39m"`;

exports[`Confirm errors for incorrectly cased input value 1`] = `
"[31mâœ–[39m [1mConfirm?[22m [90m(y/N)[39m
  [31mPlease select \\"y\\" or \\"N\\"[39m"
`;

exports[`Confirm errors for invalid input value 1`] = `
"[31mâœ–[39m [1mConfirm?[22m [90m(y/N)[39m
  [31mPlease select \\"y\\" or \\"N\\"[39m"
`;

exports[`Confirm errors with custom message 1`] = `
"[31mâœ–[39m [1mConfirm?[22m [90m(y/N)[39m
  [31mWrong value yo[39m"
`;

exports[`Confirm falsy renders success state 1`] = `"[32mâœ”[39m [1mConfirm?[22m [35mfalse[39m"`;

exports[`Confirm renders label with yes/no by default 1`] = `"[36m?[39m [1mConfirm?[22m [90m(y/N)[39m"`;

exports[`Confirm truthy renders success state 1`] = `"[32mâœ”[39m [1mConfirm?[22m [35mtrue[39m"`;
